File: packages/shared/src/membership.ts
/**
 * Membership apply policy. Signature check lives in @ollo/crypto;
 * this only decides whether a verified roster replaces local state
 * and which user ids may receive sender keys.
 */
export type MembershipDecision = "accept" | "unchanged" | "stale" | "drop";

export interface LocalMembership {
  epoch: number;
  hash: string;
}

export function planMembershipApply(args: {
  local?: LocalMembership;
  incomingEpoch: number;
  incomingHash: string;
  signatureValid: boolean;
  signerRole: string;
}): MembershipDecision {
  if (!args.signatureValid) return "drop";
  if (args.signerRole !== "admin") return "drop";
  if (args.incomingEpoch < 1 || !args.incomingHash) return "drop";
  if (!args.local) return "accept";
  if (args.incomingEpoch < args.local.epoch) return "stale";
  if (args.incomingEpoch === args.local.epoch) {
    return args.incomingHash === args.local.hash ? "unchanged" : "drop";
  }
  return "accept";
}

export function planTrustedMembers(
  signedUserIds: string[],
  serverUserIds: string[],
): { trusted: string[]; extra: string[]; missing: string[] } {
  const signed = new Set(signedUserIds.filter(Boolean));
  const server = new Set(serverUserIds.filter(Boolean));
  return {
    trusted: [...signed].filter((id) => server.has(id)),
    extra: [...server].filter((id) => !signed.has(id)),
    missing: [...signed].filter((id) => !server.has(id)),
  };
}

/** Fan-out only to the signed ∩ live intersection. Empty signed roster → nobody. */
export function planFanoutRecipients(signedUserIds: string[], serverUserIds: string[]): string[] {
  if (!signedUserIds.some(Boolean)) return [];
  return planTrustedMembers(signedUserIds, serverUserIds).trusted;
}

export function sameMembership(
  a: { userId: string; role: string }[],
  b: { userId: string; role: string }[],
): boolean {
  if (a.length !== b.length) return false;
  const as = [...a].sort((x, y) => (x.userId < y.userId ? -1 : x.userId > y.userId ? 1 : 0));
  const bs = [...b].sort((x, y) => (x.userId < y.userId ? -1 : x.userId > y.userId ? 1 : 0));
  return as.every((m, i) => m.userId === bs[i]!.userId && m.role === bs[i]!.role);
}

File: apps/android/core/crypto/src/main/java/app/ollo/crypto/Membership.kt
package app.ollo.crypto

import java.io.ByteArrayOutputStream
import java.security.MessageDigest

/**
 * Canonical membership encoding + apply policy. Must stay in lockstep with
 * `packages/crypto/src/membership.ts` and `packages/shared/src/membership.ts`.
 * Native sign/verify uses platform Ed25519 later; the hash is SHA-256.
 */
object Membership {
    const val DOMAIN = "ollo-membership-v1"

    data class Member(val userId: String, val role: String)

    data class Local(val epoch: Int, val hash: String)

    enum class Decision { Accept, Unchanged, Stale, Drop }

    fun canonicalize(members: List<Member>): List<Member> {
        val sorted = members.sortedBy { it.userId }
        val seen = HashSet<String>()
        val out = ArrayList<Member>()
        for (m in sorted) {
            if (m.userId.isEmpty()) throw IllegalArgumentException("invalid membership row")
            if (m.role != "admin" && m.role != "moderator" && m.role != "member") {
                throw IllegalArgumentException("invalid membership row")
            }
            if (!seen.add(m.userId)) throw IllegalArgumentException("duplicate membership row")
            out.add(m)
        }
        if (out.isEmpty()) throw IllegalArgumentException("empty membership")
        return out
    }

    fun encode(groupId: String, epoch: Int, members: List<Member>): ByteArray {
        if (groupId.isEmpty() || epoch < 1) throw IllegalArgumentException("invalid membership statement")
        val rows = canonicalize(members)
        val out = ByteArrayOutputStream()
        out.write(DOMAIN.toByteArray(Charsets.UTF_8))
        out.write(0)
        out.write(groupId.toByteArray(Charsets.UTF_8))
        out.write(0)
        out.write(epoch.toString().toByteArray(Charsets.UTF_8))
        out.write(0)
        for (m in rows) {
            out.write(m.userId.toByteArray(Charsets.UTF_8))
            out.write(0)
            out.write(m.role.toByteArray(Charsets.UTF_8))
            out.write(0)
        }
        return out.toByteArray()
    }

    fun hash(groupId: String, epoch: Int, members: List<Member>): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(encode(groupId, epoch, members))
        return digest.joinToString("") { b -> "%02x".format(b.toInt() and 0xff) }
    }

    fun planApply(
        local: Local?,
        incomingEpoch: Int,
        incomingHash: String,
        signatureValid: Boolean,
        signerRole: String,
    ): Decision {
        if (!signatureValid) return Decision.Drop
        if (signerRole != "admin") return Decision.Drop
        if (incomingEpoch < 1 || incomingHash.isEmpty()) return Decision.Drop
        if (local == null) return Decision.Accept
        if (incomingEpoch < local.epoch) return Decision.Stale
        if (incomingEpoch == local.epoch) {
            return if (incomingHash == local.hash) Decision.Unchanged else Decision.Drop
        }
        return Decision.Accept
    }

    fun trustedMembers(signedUserIds: List<String>, serverUserIds: List<String>): Triple<List<String>, List<String>, List<String>> {
        val signed = signedUserIds.filter { it.isNotEmpty() }.toSet()
        val server = serverUserIds.filter { it.isNotEmpty() }.toSet()
        val trusted = signed.filter { it in server }
        val extra = server.filter { it !in signed }
        val missing = signed.filter { it !in server }
        return Triple(trusted, extra, missing)
    }
}
